embassy-stm32/src/dma/ringbuffer.rs

```diff
@@ -54,151 +53,130 @@ pub trait DmaCtrl {
-        }
+        Self { dma_buf, start: 0 }
     }
 
     /// Reset the ring buffer to its initial state
     pub fn clear(&mut self, mut dma: impl DmaCtrl) {
-        self.first = 0;
-        self.ndtr = self.dma_buf.len();
+        self.start = 0;
         dma.reset_complete_count();
     }
 
-    /// The buffer end position
-    fn end(&self) -> usize {
-        self.dma_buf.len() - self.ndtr
-    }
-
-    /// Returns whether the buffer is empty
-    pub fn is_empty(&self) -> bool {
-        self.first == self.end()
+    /// The capacity of the ringbuffer
+    pub const fn len(&self) -> usize {
+        self.dma_buf.len()
     }
 
-    /// The current number of bytes in the buffer
-    /// This may change at any time if dma is currently active
-    pub fn len(&self) -> usize {
-        // Read out a stable end (the dma periheral can change it at anytime)
-        let end = self.end();
-        if self.first <= end {
-            // No wrap
-            end - self.first
-        } else {
-            self.dma_buf.len() - self.first + end
-        }
+    fn pos(&self, remaining_transfers: usize) -> usize {
+        self.len() - remaining_transfers
     }
 
     /// Read bytes from the ring buffer
+    /// Return a tuple of the length read and the length remaining in the buffer
+    /// If not all of the bytes were read, then there will be some bytes in the buffer remaining
+    /// The length remaining is the capacity, ring_buf.len(), less the bytes remaining after the read
     /// OverrunError is returned if the portion to be read was overwritten by the DMA controller.
-    pub fn read(&mut self, mut dma: impl DmaCtrl, buf: &mut [W]) -> Result<usize, OverrunError> {
-        let end = self.end();
+    pub fn read(&mut self, mut dma: impl DmaCtrl, buf: &mut [W]) -> Result<(usize, usize), OverrunError> {
+        /*
+            This algorithm is optimistic: we assume we haven't overrun more than a full buffer and then check
+            after we've done our work to see we have. This is because on stm32, an interrupt is not guaranteed
+            to fire in the same clock cycle that a register is read, so checking get_complete_count early does
+            not yield relevant information.
+
+            Therefore, the only variable we really need to know is ndtr. If the dma has overrun by more than a full
+            buffer, we will do a bit more work than we have to, but algorithms should not be optimized for error
+            conditions.
+
+            After we've done our work, we confirm that we haven't overrun more than a full buffer, and also that
+            the dma has not overrun within the data we could have copied. We check the data we could have copied
+            rather than the data we actually copied because it costs nothing and confirms an error condition
+            earlier.
+        */
+        let end = self.pos(dma.get_remaining_transfers());
+        if self.start == end && dma.get_complete_count() == 0 {
```

Comment:
This is a distinction between my implementation and the current implementation. In the event that the DMA has filled the buffer completely and then stopped (unlikely, but possible), my implementation can recover all of the data from the ringbuffer, whereas I think the current implementation would return overflow.

Supposed the ringbuffer wrapped all the way around, and the dma stopped, but the buffer was never read. In that case, complete_count would be `==1` because the buffer had wrapped, but all data would still be valid (end slices after; start slices before). The race is still solve because we always check again after we've read data.

Reply:
That is a good point. Thanks for clarifying this!.